### PR TITLE
feat: Add method to run command with terminal widget

### PR DESCRIPTION
**Pull request by lstocchi**
A bit of context. I had to use this command instead of the usual one because it doesn't seem working with the knative function cli. My terminal is stuck and nothing is printed on. I saw that the `createLocalShellWidget` solves our problem and i created a method for it.

### CODE REVIEW

File: src/main/java/com/redhat/devtools/intellij/common/utils/ExecHelper.java [hunk 1]
```diff
@@ -498,6 +500,36 @@ public static void executeWithTerminal(Project project, String title, String...
...
+      method[0] = TerminalView.class.getMethod("createLocalShellWidget", new Class[] {String.class, String.class});
+      parameters[0] = new Object[] { workingDirectory, title };
+    } catch (NoSuchMethodException e) {
+      try {
+        method[0] = TerminalView.class.getMethod("createLocalShellWidget", new Class[] {String.class});
+        parameters[0] = new Object[] { workingDirectory };
+      } catch (NoSuchMethodException e1) {
+        throw new IOException(e1);
+      }
+    }
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/common/utils/ExecHelper.java:523]: Invoking `createLocalShellWidget` via reflection (apparently in order to catch both API variants) could be extracted to a separate method for improved clarity/isolation.

File: src/main/java/com/redhat/devtools/intellij/common/utils/ExecHelper.java [hunk 2]
```diff
@@ -498,6 +500,36 @@ public static void executeWithTerminal(Project project, String title, String...
...
+    ApplicationManager.getApplication().invokeLater(() -> {
+      try {
+        ShellTerminalWidget shellTerminalWidgets = (ShellTerminalWidget) method[0].invoke(view, parameters[0]);
+        shellTerminalWidgets.executeCommand(String.join(" ", command));
+      } catch (IllegalAccessException | InvocationTargetException | IOException ignored) {}
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/common/utils/ExecHelper.java:528]: Should log if something goes wrong that throws an Exception

File: src/main/java/com/redhat/devtools/intellij/common/utils/ExecHelper.java (hunk 1 above)
Comment (adietish) [reply src/main/java/com/redhat/devtools/intellij/common/utils/ExecHelper.java]: done this in my added commit

File: src/main/java/com/redhat/devtools/intellij/common/utils/ExecHelper.java (hunk 2 above)
Comment (adietish) [reply src/main/java/com/redhat/devtools/intellij/common/utils/ExecHelper.java]: done this in my added commit

**Review by jeffmaury — CHANGES_REQUESTED** [review]
New methods do not allow to pass environment (proxy,...). So this will not work in all case. I think we should get:
- either continue with ShellTerminalWidget but then set the env (through commands)
- work with TerminalExecutionConsole which seems a better fit for processes

**Comment by adietish** [comment]
> New methods do not allow to pass environment (proxy,...). So this will not work in all case. I think we should get:
> 
>     * either continue with ShellTerminalWidget but then set the env (through commands)

The env can be set via the `localTerminalCustomizer` extension point in the plugin that is using common. This extension point is called whenever a terminal is opened. The same approach is ex. used by PyTerminal: https://github.com/JetBrains/intellij-community/commit/5f703db7df7e0b56f4680582ec69144fed79d008#diff-e7a13dca77dfd45ba213dc20019ce39ae0e332e520de0ecb3ca32acd0f11ce54R59

**Comment by jeffmaury** [comment]
Which is a terrible design: you are using a probably application scoped extension point to pass parameters to a method !!!!!

**Comment by adietish** [comment]
> Which is a terrible design: you are using a probably application scoped extension point to pass parameters to a method !!!!!

I did not design it ;) that's what IJ provides for this sake as one can see in PyTerminal. On the other hand one could argue that being able to pass specific env vars to a method is rather weird. As in bash, those exist for all processes that are run in that shell, not just for a single programm.

**Comment by jeffmaury** [comment]
What we are looking is to launch a process and dump the output in a terminal window. The ShellTerminalWidget is designed to manage a system shell and the Python plugin extension point is here to inject the virtualenv environment. In our case, the environment is more dynamic (relative to the cluster URL) that's why TerminalExecutionConsole seems better suited

**Comment by lstocchi** [comment]
I abandoned this solution as there is no easy way to set up envs.
Setting them within the command is not acceptable for me as the command is printed to the user and it would be a mess.
The only way i found out is https://github.com/redhat-developer/intellij-common/pull/91/files#diff-d6b49f55590165bc437c8ca2b4a64cf02851cb4a094f262f23749caf94e7758cR523 but it's a bit hacky and not reliable at all because if there are two consecutive calls there is the risk the envs will be overwritten.
So either we accept this risk or we add this method for ppl who wants to only run commands that do not require setting up envs. Otherwise it is not an option.

**Comment by lstocchi** [comment]
Closing it as https://github.com/redhat-developer/intellij-common/pull/108 was chosen as the final solution

**Pull request closed without merging**